File: crates/terminal/src/commands/sysinfo.rs
use crate::service::TerminalService;
use crate::error::Result;
use crate::commands::Command;
use async_trait::async_trait;

pub struct SysInfoCommand;

#[async_trait]
impl Command for SysInfoCommand {
    fn name(&self) -> &str {
        "sysinfo"
    }

    async fn execute(&self, service: &TerminalService, args: &[&str], _stdin: Option<&str>) -> Result<(String, String, i32)> {
        if args.is_empty() {
             match service.system_service.get_current_stats().await {
                Ok(stats) => {
                    match serde_json::to_string_pretty(&stats) {
                        Ok(s) => Ok((s, "".to_string(), 0)),
                        Err(e) => Ok(("".to_string(), format!("sysinfo: serialization error: {}", e), 1))
                    }
                },
                Err(e) => Ok(("".to_string(), format!("sysinfo: failed to get stats: {}", e), 1))
            }
        } else {
            match args[0] {
                "health" => {
                    match service.system_service.health().await {
                        Ok(h) => Ok((format!("{}\n", h), "".to_string(), 0)),
                        Err(e) => Ok(("".to_string(), format!("Error getting health: {}", e), 1))
                    }
                },
                "device" => {
                    match service.system_service.get_device_info().await {
                        Ok(info) => Ok((format!("{:?}\n", info), "".to_string(), 0)),
                        Err(e) => Ok(("".to_string(), format!("Error getting device info: {}", e), 1))
                    }
                },
                "gpu" => {
                    let gpus = service.system_service.get_gpus().await;
                    Ok((format!("{:?}\n", gpus), "".to_string(), 0))
                },
                "ports" => {
                    if args.len() < 2 {
                         return Ok(("".to_string(), "sysinfo ports: requires port numbers".to_string(), 1));
                    }
                    let ports: Vec<u16> = args[1..].iter().filter_map(|p| p.parse().ok()).collect();
                    match service.system_service.check_ports(ports).await {
                        Ok(statuses) => Ok((format!("{:?}\n", statuses), "".to_string(), 0)),
                        Err(e) => Ok(("".to_string(), format!("Error checking ports: {}", e), 1))
                    }
                },
                "docker-mirrors" => {
                    if args.len() < 2 {
                         return Ok(("".to_string(), "sysinfo docker-mirrors: requires subcommand [get|set]".to_string(), 1));
                    }
                    match args[1] {
                        "get" => {
                             match service.system_service.get_docker_mirrors().await {
                                 Ok(mirrors) => Ok((format!("{:?}\n", mirrors), "".to_string(), 0)),
                                 Err(e) => Ok(("".to_string(), format!("Error getting docker mirrors: {}", e), 1))
                             }
                        },
                        "set" => {
                             if args.len() < 3 {
                                 return Ok(("".to_string(), "sysinfo docker-mirrors set: requires mirrors json array".to_string(), 1));
                             }
                             let json_str = args[2];
                             match serde_json::from_str::<Vec<serde_json::Value>>(json_str) {
                                 Ok(mirrors) => {
                                     match service.system_service.set_docker_mirrors(mirrors).await {
                                         Ok(_) => Ok(("Docker mirrors updated\n".to_string(), "".to_string(), 0)),
                                         Err(e) => Ok(("".to_string(), format!("Error setting docker mirrors: {}", e), 1))
                                     }
                                 },
                                 Err(e) => Ok(("".to_string(), format!("Invalid mirrors JSON: {}", e), 1))
                             }
                        },
                        _ => Ok(("".to_string(), format!("sysinfo docker-mirrors: unknown subcommand '{}'", args[1]), 1))
                    }
                },
                 "docker-settings" => {
                    if args.len() < 2 {
                         return Ok(("".to_string(), "sysinfo docker-settings: requires subcommand [get|set]".to_string(), 1));
                    }
                    match args[1] {
                        "get" => {
                            match service.system_service.get_docker_settings().await {
                                Ok(settings) => Ok((format!("{:?}\n", settings), "".to_string(), 0)),
                                Err(e) => Ok(("".to_string(), format!("Error getting docker settings: {}", e), 1))
                            }
                        },
                        "set" => {
                             if args.len() < 3 {
                                 return Ok(("".to_string(), "sysinfo docker-settings set: requires settings json object".to_string(), 1));
                             }
                             let json_str = args[2];
                             match serde_json::from_str::<serde_json::Value>(json_str) {
                                 Ok(settings) => {
                                     match service.system_service.set_docker_settings(settings).await {
                                         Ok(_) => Ok(("Docker settings updated\n".to_string(), "".to_string(), 0)),
                                         Err(e) => Ok(("".to_string(), format!("Error setting docker settings: {}", e), 1))
                                     }
                                 },
                                 Err(e) => Ok(("".to_string(), format!("Invalid settings JSON: {}", e), 1))
                             }
                        },
                        _ => Ok(("".to_string(), format!("sysinfo docker-settings: unknown subcommand '{}'", args[1]), 1))
                    }
                },
                "history" => {
                     let query = domain::system::StatsHistoryQuery {
                         start: None,
                         end: None,
                         limit: Some(10), // Default to last 10
                     };
                     match service.system_service.get_stats_history(query).await {
                         Ok(history) => {
                             let mut output = String::from("TIMESTAMP                    CPU(%)  MEM(%)  DISK(%)\n");
                             for stats in history {
                                 output.push_str(&format!("{:?}  {:.1}    {:.1}    {:.1}\n", 
                                     stats.created_at.unwrap_or_default(), 
                                     stats.cpu_usage, 
                                     stats.memory_usage, 
                                     stats.disk_usage
                                 ));
                             }
                             Ok((output, "".to_string(), 0))
                         },
                         Err(e) => Ok(("".to_string(), format!("Error getting stats history: {}", e), 1))
                     }
                },
                _ => Ok(("".to_string(), format!("sysinfo: unknown subcommand '{}'", args[0]), 1))
            }
        }
    }
}

File: crates/kernel/src/userland/sysinfo.js
const args = sys.args || [];
const system = sys.system;

if (args.length === 0) {
    try {
        const stats = JSON.parse(system.stats());
        print(JSON.stringify(stats, null, 2));
    } catch (e) {
        print(`sysinfo: failed to get stats: ${e}`);
    }
} else {
    const cmd = args[0];
    switch (cmd) {
        case "health":
            print(system.health());
            break;
        case "device":
            try {
                const dev = JSON.parse(system.device());
                print(JSON.stringify(dev, null, 2));
            } catch (e) {
                print(`Error getting device info: ${e}`);
            }
            break;
        case "gpu":
            try {
                const gpu = JSON.parse(system.gpu());
                print(JSON.stringify(gpu, null, 2));
            } catch (e) {
                print(`Error getting gpu info: ${e}`);
            }
            break;
        case "ports":
            if (args.length < 2) {
                print("sysinfo ports: requires port numbers");
            } else {
                const ports = args.slice(1).map(p => parseInt(p)).filter(p => !isNaN(p));
                try {
                    const res = JSON.parse(system.checkPorts(ports));
                    print(JSON.stringify(res, null, 2));
                } catch (e) {
                    print(`Error checking ports: ${e}`);
                }
            }
            break;
        case "docker-mirrors":
            if (args.length < 2) {
                print("sysinfo docker-mirrors: requires subcommand [get|set]");
            } else {
                const sub = args[1];
                if (sub === "get") {
                    try {
                        const mirrors = JSON.parse(system.getDockerMirrors());
                        print(JSON.stringify(mirrors, null, 2));
                    } catch (e) {
                        print(`Error getting docker mirrors: ${e}`);
                    }
                } else if (sub === "set") {
                     if (args.length < 3) {
                        print("sysinfo docker-mirrors set: requires mirrors json array");
                    } else {
                        const jsonStr = args[2];
                        try {
                            // Validate JSON first
                            const mirrors = JSON.parse(jsonStr);
                            if (!Array.isArray(mirrors)) throw "Mirrors must be an array";
                            system.setDockerMirrors(jsonStr);
                            print("Docker mirrors updated");
                        } catch (e) {
                            print(`Error setting docker mirrors: ${e}`);
                        }
                    }
                } else {
                    print(`sysinfo docker-mirrors: unknown subcommand '${sub}'`);
                }
            }
            break;
         case "docker-settings":
            if (args.length < 2) {
                print("sysinfo docker-settings: requires subcommand [get|set]");
            } else {
                const sub = args[1];
                if (sub === "get") {
                    try {
                        const settings = JSON.parse(system.getDockerSettings());
                        print(JSON.stringify(settings, null, 2));
                    } catch (e) {
                        print(`Error getting docker settings: ${e}`);
                    }
                } else if (sub === "set") {
                     if (args.length < 3) {
                        print("sysinfo docker-settings set: requires settings json object");
                    } else {
                        const jsonStr = args[2];
                        try {
                            system.setDockerSettings(jsonStr);
                            print("Docker settings updated");
                        } catch (e) {
                            print(`Error setting docker settings: ${e}`);
                        }
                    }
                } else {
                    print(`sysinfo docker-settings: unknown subcommand '${sub}'`);
                }
            }
            break;
        default:
            print(`sysinfo: unknown subcommand '${cmd}'`);
    }
}
